students/TimPauley/session10/database.py

```diff
@@ -0,0 +1,182 @@
+                row_dict[column] = row[index]
+
+            dict_list.appendTime(row_dict)
+
+        return dict_list
+
+
+@timing
+def import_data(directory_name, product_file, customer_file, rentals_file):
+    """
+Functions to import files
+	1. Products
+	2. Customers
+	3. Rentals
+
+    """
+    product_errors = 0
+    customer_errors = 0
+    rental_errors = 0
+    try:
+        products = db["products"]
+        products.insert_many(import_csv(product_file))
+    except ImportError:
+        product_errors += 1
+    try:
+        customers = db["customers"]
+        customers.insert_many(import_csv(customer_file))
+    except ImportError:
+        customer_errors += 1
+    try:
+        rentals = db["rentals"]
+        rentals.insert_many(import_csv(rentals_file))
+    except ImportError:
+        rental_errors += 1
+
+    count_records = (db.products.find().count(), db.customers.find().count(), db.rentals.find().count())
+    count_errors = (product_errors, customer_errors, rental_errors)
+
+    return count_records, count_errors
+
+
+@timing
+def show_available_products(db):
+    """
+    """
+    available_products = {}
+    for product_id in db.products.find():
+        product_dict = {"description": product_id["description"],
+                    "product_type": product_id["product_type"],
+                    "quantity_available": product_id["quantity_available"]}
+        if product_id["quantity_available"] != "0":
+            available_products[product_id["product_id"]] = product_dict
+            continue
+        else:
+            continue
+
+    return available_products
+
+
+@timing
+def show_rentals(db, product_id):
+    """
+    this function shows rentals 
+    """
+    rental_users_dict = {}
+    for rental in db.rentals.find():
+        if rental["product_id"] == product_id:
+            customer_id = rental["user_id"]
+
+            customer_record = db.customers.find_one({"user_id": customer_id})
+
+            rental_users = {"name": customer_record["name"],
+                            "address": customer_record["address"],
+                            "phone_number": customer_record["phone_number"],
+                            "email": customer_record["email"]}
+            rental_users_dict[customer_id] = rental_users
+            continue
+        else:
+            continue
+
+    return rental_users_dict
+
+
+if __name__ == "__main__":
+    mongo = MongoDBConnection()
+
+    with mongo:
+        # here is where we use the mongo connection
+        db = mongo.connection.media
+
+
+        import_data("", "products.csv", "customers.csv", "rentals.csv")
+
+        show_available_products(db)
+
+        show_rentals(db, "prd005")
+
+        # By dropping the tables we clear the for next time
+        drop_prompt = input("Your tables have been dropped")
+        if drop_prompt.upper() == 'Y':
+            db.products.drop()
+            db.customers.drop()
+            db.rentals.drop()
+
```

Comment:
function decorator is probably not what he had in mind, but it is meta, and you are more likely to be using it in the future than the other meta techniques, so go for it. :)